brave/README.md

```diff
@@ -772,3 +772,68 @@ Propagating a trace context instead of a span is a right fit for several reasons
 ### Public namespace
 Brave 4's public namespace is more defensive that the past, using a package
 accessor design from [OkHttp](https://github.com/square/okhttp).
+
+### CorrelationFields Api
+
+The CorrelationFields functionality is similar to OpenTracing (OT) Baggage, and
+dissimilar to Census tags. This is a legacy preference, allowing Sleuth users
+(which has a baggage api similar to OT) to transition to Brave. This design
+implies string based correlation fields are stored in a span or a trace context
+for read-back, whereas census tags are immutable scoped variables. OT, Census
+and Brave all decouple thread scoping, header injection and extraction from the
+apis that mutate propagated fields.
+
+Here's an example of using OpenTracing baggage:
+```java
+// mutate the span's baggage adding the method
+span.setBaggageItem("method", fullMethodName);
+
+// set the span in scope
+try (Scope scope = tracer.scopeManager().activate(span, false)) {
+  // If downstream code call tracer.activeSpan(), they'll see the above span.
+  // It can then read span.getBaggageItem("method")
+  //               or span.context().baggageItems()
+}
+// Repeating the same block might have different results as downstream code
+// or other threads can overwrite the key name "method".
+```
+
+Here's an example of using Census tagging:
+```java
+// Tags are encouraged to be constants
+static final TagKey RPC_METHOD = TagKey.create("method");
+...
+
+// make a copy of the parent tag context, with the method tag
+TagContext withMethod = tagger.toBuilder(parentCtx)
+                              .put(RPC_METHOD, TagValue.create(fullMethodName)).build();
+
+// set the tag context in scope
+try (Scope scope = tagger.withTagContext(withMethod)) {
+  // If downstream code call tagger.getCurrentTagContext(), they'll see the above tags.
+  // However, as of v0.10, TagContext.getIterator() is hidden, so only census impls can read them!
+}
+// Repeating the same block has the exact same result as tag contexts are immutable.
+```
+
+The OT-centric design has a few implications, notably shared mutable state:
+
+* String keys and values are easy to use, but don't imply up-front validation.
+  * Census tags are types, encouraging prevalidation and use as constants.
+* Sharing mutable data across threads can lead to inconsistent reads of fields.
+  * For example, child threads affect the same state their parent reads.
+  * Immutable tag contexts eliminate this
+* A tracer api is more complex than a tagger api, and has an ordering concern.
+  * To contribute tags, you need to have tracing start before metrics.
```

Comment:
I'm not entirely sure I get the point of bringing up metrics here. Ideally only a subset of span tags (the ones ensured to be low cardinality + `http.template`) and none of these seem to be related to correlation fields. What's the link I'm missing here?

Reply:
yeah wording was a bit crap. (tags are the census name for correlation fields).

The jist is that if the correlation api is separate, it can scope fields before or even if tracing doesn't exist. One use case is to use the same correlation field in metrics and tracing (and further down the line, such as country-code)

To have metrics re-use correlation fields from tracing, for example, it would likely need to start after tracing started (or at the same time). Similar problem exists in logging. If you want logs to contain a correlation field, that field should be scoped as early as possible. Otherwise, you see request handling without and MDC value until sometime later when tracing kicks in.

The idea is that a tagging api could be made at the head of a request, and be simpler or at least less dangerously, ensuring that whether it is tracing metrics or logging (or something else), they all read exactly the same values, with no ordering interdependency apart from tagger.

What I'm calling out here is that if we assign that responsibility to brave (for convenience because parsing, inject extract propagation etc is already handled), there is at least an interdependency problem, which is most easily sorted if trace starts before other subsystems on a request, or at least as early as possible.

Now.. if I did clear up anything (mildly possible), could you help me revise it into a shorter bullet or two? :)